Source/WebKit/Platform/CoroutineUtilities.h

```diff
@@ -0,0 +1,180 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/CompletionHandler.h>
+
+#if __has_include(<coroutine>)
+#include <coroutine>
+#else
+#include <experimental/coroutine>
+namespace std {
+using std::experimental::coroutine_handle;
+using std::experimental::suspend_never;
+using std::experimental::suspend_always;
+}
+#endif
+
+namespace WebKit {
+
+template<typename PromiseType>
+class CoroutineHandle {
+public:
+    CoroutineHandle() = default;
+    CoroutineHandle(std::coroutine_handle<PromiseType>&& handle)
+        : m_handle(std::exchange(handle, nullptr)) { }
+    CoroutineHandle(CoroutineHandle&& other)
+        : m_handle(std::exchange(other.m_handle, nullptr)) { }
+    CoroutineHandle& operator=(CoroutineHandle&& other)
+    {
+        m_handle = std::exchange(other.m_handle, nullptr);
+        return *this;
+    }
+    CoroutineHandle(const CoroutineHandle&) = delete;
+    CoroutineHandle& operator=(const CoroutineHandle&) = delete;
+    ~CoroutineHandle()
+    {
+        if (m_handle)
+            m_handle.destroy();
+    }
+
+    std::coroutine_handle<PromiseType> handle() const { return m_handle; }
+private:
+    std::coroutine_handle<PromiseType> m_handle;
+};
+
+// Name based on https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1056r1.html
+template<typename T>
+class Lazy {
+public:
+    class PromiseBase {
+    public:
+        struct final_awaitable {
+            bool await_ready() const noexcept { return false; }
+            template<typename Promise>
+            std::coroutine_handle<> await_suspend(std::coroutine_handle<Promise> coroutine) noexcept { return coroutine.promise().handle(); }
+            void await_resume() noexcept { }
+        };
+        std::suspend_always initial_suspend() { return { }; }
+        final_awaitable final_suspend() noexcept { return { }; }
+        template<typename Promise>
+        void unhandled_exception() { }
+        void setHandle(std::coroutine_handle<> handle) { m_handle = handle; }
+        std::coroutine_handle<> handle() { return m_handle; }
+    private:
+        std::coroutine_handle<> m_handle;
+    };
+    template<typename U> class Promise final : public PromiseBase {
+    public:
+        Lazy<U> get_return_object() { return Lazy<U> { std::coroutine_handle<Promise>::from_promise(*this) }; }
+        void return_value(U&& value) { m_value = WTFMove(value); }
+        U result() && { return std::exchange(m_value, { }); }
+    private:
+        U m_value;
+    };
+    template<> class Promise<void> : public PromiseBase {
```

Comment:
I believe the GTK/WPE bots are unhappy because of a GCC bug, and you'll have to use this workaround from the bug report: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85282#c17